### DISCUSSION TITLE
Keeping Track of Spend UTXOs

### DISCUSSION THREAD

**Discussion by bakon11** [CardanoSolutions/kupo#19] [Archives]
Currently when I query Kupo for a match using an Address it will return every UTXO associated with that address, whether that UTXO has been spend or not.

I am not sure if Kupo has a mechanism to keep track of spend UTXOs or not. 

If there is no such mechanism would that be something that can be build using Ogmios or Oura to keep track when a UTXO was used as an input and then just prune the Kupo sqlite db of it?

Thanks.

**Comment by KtorZ** [comment]
That's a feature / options I was considering to add, waiting for a someone to have a use-case for it.

At the moment, it's pretty easy to check whether a UTxO still exist using Ogmios and an output reference. I had in mind though to allow through a flag, to automatically prune the history in Kupo to only keep the edge of the UTxO graph.

Though, may I ask, if you need such a feature, what are you missing from the UTxO query from Ogmios / cardano-node?

**Reply by bakon11** [reply]
Absolutely,

So currently we query our smart contract address to check its current UTXOs and read their Datums for information. It's not 100% implemented since there is a few things that we can't really utilize until the Vasil hf in June which is ok.

We also have a wallet build in utilizing the Cardano serialization library. And we query address UTXOs to pass for TX creation.

With the current system querying UTXOs by address(Ogmios or CLI) it only returns unspent UTXOs.

Maybe this feature doesn't have to prune the UTXO from the DB but add a `spent` flag to it.

And add a extra param to the `Delete Matches` end point that specifies remove only UTXO flagged spend although that can be easily done on our software side too.

**Comment by satran004** [comment]
I have similar requirement in cardano-client Java library where utxos for an address are fetched from Kupo to build transaction.

I understand that it's possible to verify if an output is spent or not through Ogmios. But I am wondering if we can avoid that extra call(s).

For example, I am getting around ~2942 outputs from the following testnet address. But only 3 outputs are unspent.
As @bakon11 suggested, if it can be handled  through a spent/unspent flag in Kupo, few extra calls to Ogmios can avoided.  

addr_test1qzx9hu8j4ah3auytk0mwcupd69hpc52t0cw39a65ndrah86djs784u92a3m5w475w3w35tyd6v3qumkze80j8a6h5tuqq5xe8y
